include/eve/detail/function/simd/x86/slide_left.hpp

```diff
@@ -0,0 +1,147 @@
+          auto const w = bit_cast(v, as_<f_t>{});
+
+          if constexpr(Shift == 1)
+          {
+            return bit_cast(f_t(_mm256_shuffle_pd(w,_mm256_permute2f128_pd(w,w,0x81),0x5)), as(v));
+          }
+          else if constexpr(Shift == 2)
+          {
+            return bit_cast(f_t(_mm256_permute2f128_pd(w, w, 0x81)), as(v));
+          }
+          else if constexpr(Shift == 3)
+          {
+            auto const s0 = _mm256_permute2f128_pd(w,w,0x81);
+            return bit_cast(f_t(_mm256_shuffle_pd(s0,f_t(0),1)), as(v));
+          }
+        }
+        else if constexpr(sizeof(T) == 4)
+        {
+          // Seven optimal shuffles for sliding left by 1->7 doctors want you to not know about
+          using f_t = as_floating_point_t<wide<T,N,ABI>>;
+          auto const w = bit_cast(v, as_<f_t>{});
+
+          if constexpr(Shift == 1)
+          {
+            auto const s0 = _mm256_blend_ps(w, _mm256_permute2f128_ps(w,w,0x81), 0x11);
+            return bit_cast(f_t(_mm256_permute_ps(s0, 0x39)), as(v));
+          }
+          else if constexpr(Shift == 2)
+          {
+            auto const s0 = _mm256_permute2f128_ps(w, w, 0x81);
+            return bit_cast(f_t(_mm256_shuffle_ps(w, s0, 0x4e)), as(v));
+          }
+          else if constexpr(Shift == 3)
+          {
+            auto const s0 = _mm256_blend_ps(w, _mm256_permute2f128_ps(w,w,0x81), 0x77);
+            return bit_cast(f_t(_mm256_permute_ps( s0, 0x93)), as(v));
+          }
+          else if constexpr(Shift == 4)
+          {
+            return bit_cast(f_t( _mm256_permute2f128_ps(w,w,0x81)),as(v));
+          }
+          else if constexpr(Shift == 5)
+          {
+            auto const s0 = _mm256_blend_ps(f_t(0), _mm256_permute2f128_ps(w,w,0x81), 0xe);
+            return bit_cast(f_t(_mm256_permute_ps(s0, 0x39)),as(v));
+          }
+          else if constexpr(Shift == 6)
+          {
+            auto const s0 = _mm256_permute2f128_ps(w,w,0x81 );
+            return bit_cast(f_t( _mm256_shuffle_ps(s0, f_t(0), 0xe)), as(v));
+          }
+          else if constexpr(Shift == 7)
+          {
+            auto const s0 = _mm256_blend_ps(f_t(0), _mm256_permute2f128_ps(w,w,0x81), 0x8);
+            return bit_cast(f_t(_mm256_permute_ps(s0, 0x93)), as(v));
+          }
+        }
+        else
+        {
+          // Small integers can be done with partial slide after slicing
+          auto[l,h] = slice(v);
+
+          // Slide higher parts as normal
+          auto h0 = slide_left(h, index<(Shift>(N::value/2)) ? N::value/2 : Shift>);
+
+          // Slide lower parts using _mm_alignr_epi8
+          constexpr auto sz = sizeof(T)* Shift;
+          using byte_t = typename wide<T,N,ABI>::template rebind<std::uint8_t,fixed<16>>;
+
+          byte_t bytes = _mm_alignr_epi8(bit_cast(h,as_<byte_t>{}), bit_cast(l,as_<byte_t>{}), sz);
+
+          // Shift everything in place
+          return wide<T,N,ABI>{bit_cast(bytes, as(h0)),h0};
+        }
```

Comment:
For avx2 consider: 

```
template <std::size_t byte_shift>
__m256i alignr256_emulated(__m256i a, __m256i b) {
  __m256i b1_a0 = _mm256_permute2x128_si256(a, b, 0x03);
  if constexpr (byte_shift == 16)
    return b1_a0;
  else if constexpr (byte_shift < 16)
    return _mm256_alignr_epi8(b1_a0, b, byte_shift);
  else
    return _mm256_alignr_epi8(a, b1_a0, byte_shift - 16);
}

wide<T,N,ABI> zeroes(0);
alignr256_emulated<-Shift * sizeof(T)>(v, zeroes);
```

So setzero + shuffle + _mm256_alignr_epi8 - 3 instructions (that if your algorithm doesn't have a zero from somewhere already)

Reply:
Will test this

Reply:
Better codegen, adopted for AVX2